pinot-core/src/main/java/org/apache/pinot/core/accounting/WorkloadAggregator.java

```diff
@@ -0,0 +1,182 @@
+  private final ConcurrentHashMap<String, Long> _concurrentTaskCPUStatsAggregator = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Long> _concurrentTaskMemStatsAggregator = new ConcurrentHashMap<>();
+
+  private final int _sleepTimeMs;
+  private final boolean _enableEnforcement;
+
+  WorkloadBudgetManager _workloadBudgetManager;
+
+  // Aggregation state
+  Map<String, List<CPUMemThreadLevelAccountingObjects.ThreadEntry>> _allWorkloads = new HashMap<>();
+  HashMap<String, Long> _finishedTaskCPUCost = new HashMap<>();
+  HashMap<String, Long> _finishedTaskMemCost = new HashMap<>();
+
+  public WorkloadAggregator(boolean isThreadCPUSamplingEnabled, boolean isThreadMemSamplingEnabled,
+      PinotConfiguration config, InstanceType instanceType, String instanceId) {
+    _isThreadCPUSamplingEnabled = isThreadCPUSamplingEnabled;
+    _isThreadMemorySamplingEnabled = isThreadMemSamplingEnabled;
+    _config = config;
+    _instanceType = instanceType;
+    _instanceId = instanceId;
+
+    _workloadBudgetManager = Tracing.ThreadAccountantOps.getWorkloadBudgetManager();
+    _sleepTimeMs = _config.getProperty(CommonConstants.Accounting.CONFIG_OF_WORKLOAD_SLEEP_TIME_MS,
+        CommonConstants.Accounting.DEFAULT_WORKLOAD_SLEEP_TIME_MS);
+    _enableEnforcement = _config.getProperty(CommonConstants.Accounting.CONFIG_OF_WORKLOAD_ENABLE_COST_ENFORCEMENT,
+        CommonConstants.Accounting.DEFAULT_WORKLOAD_ENABLE_COST_ENFORCEMENT);
+
+    LOGGER.info("WorkloadAggregator initialized with isThreadCPUSamplingEnabled: {}, isThreadMemorySamplingEnabled: {}",
+        _isThreadCPUSamplingEnabled, _isThreadMemorySamplingEnabled);
+  }
+
+  @Override
+  public void updateConcurrentCpuUsage(String workload, long cpuUsageNS) {
+    _concurrentTaskCPUStatsAggregator.compute(workload,
+        (key, value) -> (value == null) ? cpuUsageNS : (value + cpuUsageNS));
+  }
+
+  @Override
+  public void updateConcurrentMemUsage(String workload, long memBytes) {
+    _concurrentTaskMemStatsAggregator.compute(workload,
+        (key, value) -> (value == null) ? memBytes : (value + memBytes));
+  }
+
+  @Override
+  public void cleanUpPostAggregation() {
+    // No-op.
+  }
+
+  @Override
+  public int getAggregationSleepTimeMs() {
+    return _sleepTimeMs;
+  }
+
+  @Override
+  public void preAggregate(List<CPUMemThreadLevelAccountingObjects.ThreadEntry> anchorThreadEntries) {
+    _allWorkloads.clear();
+    _finishedTaskCPUCost.clear();
+    _finishedTaskMemCost.clear();
+  }
+
+  @Override
+  public void aggregate(Thread thread, CPUMemThreadLevelAccountingObjects.ThreadEntry threadEntry) {
+    CPUMemThreadLevelAccountingObjects.TaskEntry currentQueryTask = threadEntry.getCurrentThreadTaskStatus();
+    if (currentQueryTask == null) {
+      // This means that task has finished or the task doesn't have a workload.
+      return;
+    }
+
+    String currentTaskWorkload = currentQueryTask.getWorkloadName();
```

Comment:
Workload can be null ? Its tagged as `@Nullable` in `CPUMemThreadLevelAccountantObjects`